src/js/plugin.js

```diff
@@ -0,0 +1,436 @@
+  return instance;
+};
+
+/**
+ * Takes a plugin sub-class and returns a factory function for generating
+ * instances of it.
+ *
+ * This factory function will replace itself with an instance of the requested
+ * sub-class of Plugin.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of the plugin.
+ *
+ * @param  {Plugin} PluginSubClass
+ *         The class-based plugin.
+ *
+ * @return {Function}
+ *         A factory function for the plugin sub-class.
+ */
+const createPluginFactory = (name, PluginSubClass) => {
+
+  // Add a `name` property to the plugin prototype so that each plugin can
+  // refer to itself by name.
+  PluginSubClass.prototype.name = name;
+
+  return function(...args) {
+    this[name] = new PluginSubClass(...[this, ...args]);
+    return this[name];
+  };
+};
+
+class Plugin {
+
+  /**
+   * Plugin constructor.
+   *
+   * Subclasses should call `super` to ensure plugins are properly initialized.
+   *
+   * @param {Player} player
+   *        A Video.js player instance.
+   */
+  constructor(player) {
+    this.player = player;
+
+    if (this.constructor === Plugin) {
+      throw new Error('Plugin must be sub-classed; not directly instantiated.');
+    }
+
+    // Make this object evented, but remove the added `trigger` method so we
+    // use the prototype version instead.
+    evented(this);
+    delete this.trigger;
+
+    stateful(this, this.constructor.defaultState);
+    markPluginAsActive(player, this.name);
+
+    // Bind all plugin prototype methods to this object.
+    Object.keys(Plugin.prototype).forEach(k => {
+      if (typeof this[k] === 'function') {
+        this[k] = Fn.bind(this, this[k]);
+      }
+    });
+
+    // If the player is disposed, dispose the plugin.
+    player.on('dispose', this.dispose);
+
+    /**
+     * Signals that a plugin (both basic and class-based) has just been set up
+     * on a player.
+     *
+     * In all cases, an object containing the following properties is passed as a
+     * second argument to event listeners:
+     *
+     * - `name`: The name of the plugin that was set up.
+     * - `plugin`: The raw plugin function.
+     * - `instance`: For class-based plugins, the instance of the plugin sub-class,
+     *   but, for basic plugins, the return value of the plugin invocation.
+     *
+     * @event pluginsetup
+     * @memberof Player
+     * @instance
+     */
+    player.trigger('pluginsetup', this.getEventHash_());
+  }
+
+  /**
+   * Each event triggered by plugins includes a hash of additional data with
+   * conventional properties.
+   *
+   * This returns that object or mutates an existing hash.
+   *
+   * @private
+   * @param  {Object} [hash={}]
+   *         An object to be used as event an event hash.
+   *
+   * @return {Object}
+   *         The event hash object with, at least, the following properties:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
```

Comment:
Use the custom typedef that should be defined for use in the `Player#pluginsetup` event.